File: maximum-subarray/maximum-subarray.go
// https://leetcode.com/problems/maximum-subarray/
// 106 ms 8.8 MB

package main

func maxSubArray(nums []int) int {
	sum := 0
	max := nums[0]
	for i := range nums {
		sum += nums[i]
		if sum > max {
			max = sum
		}
		if sum < 0 {
			sum = 0
		}
	}
	return max
}

File: maximum-subarray/maximum-subarray.c
// https://leetcode.com/problems/maximum-subarray/
// 110 ms 12.3 MB

// time: O(n)
int maxSubArray(int* nums, int numsSize) {
    int i;
    int max = nums[0];
    int sum = 0;
    for (i = 0; i < numsSize; i++) {
        sum += nums[i];
        if (sum > max)
            max = sum;
        if (sum < 0)
            sum = 0;
    }

    return max;
}